### DISCUSSION TITLE
[vcpkg upgrade] change behavior of vcpkg upgrade --no-dry-run

### DISCUSSION THREAD

**Discussion by Neumann-A** [microsoft/vcpkg#17922] [Ideas]
currentlly vcpkg upgrade --no-dry-run first removes all outdated packages and then tries to Rebuild them.

The problem: If one of the packages fails to build the information of all removed packages is lost.

Solution: Either have a cache file for upgrades or use a different approach for upgrade like: remove package; rebuild package. Another solution would be to not stop the whole upgrade process if one of the builds fails.

**Comment by Neumann-A** [comment]
ok after reading the source I found the option --keep-going. 
Maybe this should be the default? 
If defaultet rename the option to --stop-at-failure to also have the old behavior

**Comment by Neumann-A** [comment]
and maybe add a help dialog for it ....vcpkg help upgrade does not work or list the available commands

**Comment by degski** [comment]
@Neumann-A I agree with you, `--keep-going` should be the default. I also think `--no-dry-run` should be changed to `--dry-run` (it's just too irritating to have to type that every time (with the obvious thick fingered typos), while usually you just want to update (Inshallah).

**Comment by Neumann-A** [comment]
@degski since there is also the `update ` command i disagree on the change of the command `--no-dry-run`. `upate` and `upgrade` are really easy to mix up.

**Comment by NulAsh** [comment]
https://github.com/Microsoft/vcpkg/issues/3864

**Comment by NulAsh** [comment]
I see two options to solve this problem:

1. Change order of actions when upgrading. Download first package, compile it, remove previous version, install. Download second package etc... In case of any problem, interrupted package will still be installed, just previous version, and everything else will be untouched.

2. Do everything as usual, but before upgrading create a file with list of packages to upgrade. In case of problems, just start "vcpkg upgrade" again, it will find this file, and continue from the point of interruption.

**Comment by MikeGitb** [comment]
Nice that the issues around this topic got cleaned up. Is there any plan to solve the issue in the near future?

**Comment by vicroms** [comment]
There have been attempts to do this in PRs but the implementation is complicated.
Our recommendation to somewhat mitigate this problem is to use [manifests](https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md) if possible.

We may look back into this in the future and that's why we're leaving the issue open.

**Comment by diablodale** [comment]
Yet nother scenario of this common failure

```
git pull
git update
  "got some new things..."
git upgrade --no-dry-run
  "delete, delete, delete, delete..."

"Detecting compiler hash for triplet x64-windows-static-md-v142-sdk10b17134...
Error: while detecting compiler information:
The log content at C:\njs\vcpkg\buildtrees\detect_compiler\stdout-x64-windows-static-md-v142-sdk10b17134.log is:
CMake Error at scripts/cmake/vcpkg_minimum_required.cmake:44 (message):
  Your vcpkg executable is from 2021-07-21 which is older than required by
  the caller of vcpkg_minimum_required(VERSION 2021-08-03).  Please
  re-acquire vcpkg by running bootstrap-vcpkg.
Call Stack (most recent call first):
  scripts/ports.cmake:83 (vcpkg_minimum_required)"

and dead. vcpkg purged packages, then did needed checks, failed, and did not recover.
Naturally, all necessary checks should be done first...before any changes are persisted on the drive.

Workaround? None. The damage is done. The packages are gone.

**Reply by zapcity** [reply]
Just had this happen to me https://github.com/microsoft/vcpkg/issues/21957 Most annoying.

**Comment by Yelnats321** [comment]
The builds commonly fail spuriously as well, maybe vcpkg can dump the list of packages we had before it deleted all of them and we can feed it back that list? I am using vcpkg personally, so having a manifest for each project is more work than I want to do, but I also want to update my packages without having to mentally note all of them and reinstall if one fails.

**Comment by MikeGitb** [comment]
> maybe vcpkg can dump the list of packages we had before it deleted all of them and we can feed it back that list?

I support that

In particular with long running upgrades, there can be a million things that go wrong. Loss of internet connection, pc goes to standby, you need to interrupt the upgrade, because you need the processing power for some high priority work, you accidentally close the powershell.... 

The fact that all of that leaves you with a broken installation witjout an easy way to recover/rollback/continue is really bad design. 

Or how about,  you automatically write the set of packages that will be upgrqded to a temporary file that gets deleted after a successful run.  Then add a vcpkg --continue/retry command that will try to install the remaining packages, if the file exists.
